server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,440 @@
+        --    `wait` is the mechanism that propagates any exception that might be thrown.
+        --     Since we do not use `wait`, if an exception is thrown inside the inner action, then it will just silently die.
+
+        -- It's important that the below action is the outer action of the `withAsync` as
+        -- we rely on the Exceptions the outer action throws for 4 reasons:
+        -- 1. For `withAsync` to kill the inner "updaterThread" thread when the outer action throws.
+        -- 2. To propagate the SubscriptionTimeout exception from the `withSubscribeTimeout`.
+        -- 3. To propagate the ConnectionException from the `receiveDataMessage` that is
+        --    thrown within the "messageHandlerThread".
+        -- 4. So that `onDisconnect` runs, see that `manageConnection` is in a `finally`.
+        ( const $
+            withSubscribeTimeout
+              client
+              (startMessageHandlerThread client)
+        )
+
+    -- Simply ignore connection errors, otherwise, Warp handles the exception
+    -- and sends a 500 response in the middle of a WebSocket connection, and
+    -- that violates the WebSocket protocol.
+    -- Note that subscribers are still properly removed by the finally below.
+
+    -- We can also ignore the websocket 'CloseRequest' exception because wai-websockets ensures
+    -- that the connection is closed when this IO action finishes, as this IO action is executing in a bracket.
+    -- https://github.com/yesodweb/wai/blob/c92d7b1993338fb0f91e0f5f34fb9678871028a0/wai-websockets/Network/Wai/Handler/WebSockets.hs#L98
+    handleConnectionError :: WS.ConnectionException -> IO ()
+    handleConnectionError _ = pure ()
+
+    -- Since we are in a bracket that closes the connection, let the action silently exit.
+    handleSubscriptionTimeout :: SubscriptionTimeout -> IO ()
+    handleSubscriptionTimeout _ = pure ()
+
+  -- Put the client in the subscription tree and keep the connection open.
+  -- Remove it when the connection is closed.
+  Exception.finally
+    (onConnect client >> manageConnection)
+    (onDisconnect client)
+    `Exception.catch` handleConnectionError
+    `Exception.catch` handleSubscriptionTimeout
+
+-- After timeout, throw the action an exception if the client hasn't subscribed.
+withSubscribeTimeout :: Client -> IO a -> IO a
+withSubscribeTimeout client action = do
+  let
+    initalSubscriptionTimeout =
+      Config.configInitialSubscriptionTimeoutMicroSeconds $
+        Core.coreConfig $
+          clientCore client
+
+    clientNoSubscribers = do
+      let subscriptions = clientSubscriptions client
+      HashMap.null <$> MVar.readMVar subscriptions
+
+  threadId <- Concurrent.myThreadId
+  Monad.void $ Concurrent.forkIO $ do
+    Concurrent.threadDelay initalSubscriptionTimeout
+    noSubscribers <- clientNoSubscribers
+    Monad.when noSubscribers (Concurrent.throwTo threadId SubscriptionTimeout)
```

Comment:
This would also kill the connection when a client subscribed and then unsubscribed again within the timeout, right?

If so, that's not desirable, we should only care about the first subscription here, just to ensure the client really has valid business to do (by means of authenticating with that first subscription request).